### PR TITLE
Add snapshot zone mount

### PR DESCRIPTION
**Pull request by weilewei**
This PR:
- modify `MKFS`
  - write an empty snapshot record in snapshot zone instead of metazone
  - write an empty superblock both in snapshot and metadata zones

- add MKFS tests
  - ensure first snapshot zone is not empty (since it has an empty snapshot) and rest of snapshot zones are empty

- Modify `Mount` for snapshot and metadata zones
  - Recover from the snapshot zone with the highest superblock sequence number. Within that zone, we find the latest snapshot record.
  - recover from the metadata zone the highest superblock sequence number. Within that zone, we find the latest metadata record.

- added `RecoverFromSnapshotZone` and `RecoverFromMetaZone` functions.
  - most of the logic are the same from `RecoverFrom` function except 1) the `break` condition is changed to `continue` and 2) delete a case (left questions for now).

This PR is not intended to add metadata and snapshot rollover functions

### CODE REVIEW

File: fs/fs_zenfs.cc [hunk 1]
```diff
@@ -823,6 +823,149 @@ Status ZenFS::DecodeFileDeletionFrom(Slice* input) {
...
+          Warn(logger_, "Could not decode complete snapshot: %s",
+               s.ToString().c_str());
+          return s;
+        }
+        // found one snapshot record, and
+        // continue to find the latest valid snapshot record
+        found_one_snapshot = true;
+        continue;
+
+      /* TODO: do we need kFileUpdate and kFileDeletion for snapshot recovery ? */
```
Comment (weilewei) [inline fs/fs_zenfs.cc]: question here @levisonchen

File: fs/fs_zenfs.cc [hunk 2]
```diff
@@ -823,6 +823,149 @@ Status ZenFS::DecodeFileDeletionFrom(Slice* input) {
...
+    if (!GetFixed32(&record, &tag)) break;
+
+    if (tag == kEndRecord) break;
+
+    if (!GetLengthPrefixedSlice(&record, &data)) {
+      return Status::Corruption("ZenFS", "No recovery record data");
+    }
+
+    switch (tag) {
+      /* TODO: for metazone recovery, if we run into kCompleteFilesSnapshot,
```
Comment (weilewei) [inline fs/fs_zenfs.cc]: question here @levisonchen

File: fs/fs_zenfs.cc [hunk 3]
```diff
@@ -894,22 +1037,142 @@ Status ZenFS::RecoverFrom(ZenMetaLog* log) {
...
+    return Status::IOError("Failed to mount filesystem due to "
+                           "snapshot recovery");
+  }
+
+  Info(logger_, "Recovered from snapshot zone: %d",
+         (int)valid_zones_snapshot[snapshot_recovered_index]->GetZoneNr());
+  snapshot_superblock_ =
+    std::move(valid_superblocks_snapshot[snapshot_recovered_index]);
+
+  // TODO: @Weile Wei think we only need to do the following steps
```
Comment (weilewei) [inline fs/fs_zenfs.cc]: question here @levisonchen

**Review by weilewei — COMMENTED** [review]
can you please take a look for the questions (commented inline) for my WIP PR? @levisonchen Thanks.

File: test/zenfs_metazone_rollover_test.cc [hunk 4]
```diff
@@ -31,35 +34,69 @@ int test_metadata_rollover() {
...
+  if (!s.ok()) {
+    fprintf(stderr, "Failed to create file system, error: %s\n",
+            s.ToString().c_str());
+    delete zenFS;
+    return 1;
   }
-#endif
 
+  /* verify if the fist snapshot zone has an empty snapshot record */
+  assert(!snapshot_zones[0].IsEmpty());
```
Comment (weilewei) [inline test/zenfs_metazone_rollover_test.cc]: also question here: how do we test if we have a valid and empty snapshot in this snapshot zone.

File: fs/fs_zenfs.cc [hunk 5]
```diff
@@ -1016,8 +1155,258 @@ Status ZenFS::Mount(bool readonly) {
...
+     If that fails go to the previous as we might have crashed when rolling
+     metadata zone.
+  */
+  for (const auto& sm : seq_map_metadata) {
+    uint32_t i = sm.second;
+    std::string scratch;
+    std::unique_ptr<ZenMetaLog> log = std::move(valid_logs_metadata[i]);
+
+    s = RecoverFromMetaZone(log.get());
+    if (!s.ok()) {
```
Comment (weilewei) [inline fs/fs_zenfs.cc:1098]: Question: in MKFSV2, we don't write an empty metadata record in the metadata zone. However, in `RecoverFromMetaZone`, we only have two valid states `kFileUpdate` and `kFileDeletion` to find a non-empty metadata. So from empty or formatted metazones, this condition `s.ok()` is always false. Thus we can't initiate `meta_log_`, `metadata_recovered_index`, `metadata_recovery_ok`, or reach lines 1356 to 1359. 

Shall we add an empty metadata record when formatting metadata zone? This will add an another state, for example `kEmptyMetadata` to `ZenFSTag` and add some other helper functions like `PersistEmptyMetadata`. 
@levisonchen

File: fs/fs_zenfs.h [hunk 6]
```diff
@@ -125,9 +125,10 @@ class ZenFS : public FileSystemWrapper {
 
   Zone* cur_meta_zone_ = nullptr;
   std::unique_ptr<ZenMetaLog> meta_log_;
-  std::unique_ptr<ZenMetaLog> meta_snapshot_log_;
+  std::unique_ptr<ZenMetaLog> snapshot_log_;
```
Comment (royguo) [inline fs/fs_zenfs.h:128]: I suggest we use `op_log_` and `snapshot_log_` to distinguish these two variables.

File: fs/fs_zenfs.h [hunk 7]
```diff
@@ -125,9 +125,10 @@ class ZenFS : public FileSystemWrapper {
...
   std::mutex metadata_sync_mtx_;
-  std::unique_ptr<Superblock> superblock_;
+  std::unique_ptr<Superblock> snapshot_superblock_;
+  std::unique_ptr<Superblock> metadata_superblock_;
```
Comment (royguo) [inline fs/fs_zenfs.h]: Same as the previous comment, we may use `snapshot_super_block_` and `op_super_block` here

File: fs/fs_zenfs.h [hunk 8]
```diff
@@ -165,18 +166,21 @@ class ZenFS : public FileSystemWrapper {
   void EncodeFileDeletionTo(ZoneFile* zoneFile, std::string* output);
 
   Status DecodeSnapshotFrom(Slice* input);
+  Status DecodeSnapshotFromAndNoCacheFiles(Slice* input);
```
Comment (royguo) [inline fs/fs_zenfs.h]: For new function declaration, we should put a comment to explain its functionality (if it is not self-explanatory)

File: fs/fs_zenfs.h [hunk 9]
```diff
@@ -165,18 +166,21 @@ class ZenFS : public FileSystemWrapper {
...
   Status DecodeFileUpdateFrom(Slice* slice);
   Status DecodeFileDeletionFrom(Slice* slice);
 
   Status RecoverFrom(ZenMetaLog* log);
+  Status RecoverFromSnapshotZone(ZenMetaLog* log);
+  Status RecoverFromMetaZone(ZenMetaLog* log);
```
Comment (royguo) [inline fs/fs_zenfs.h]: Did you mean to recover from `op_log_zone`? or both `snapshot & op` zones

File: test/zenfs_metazone_rollover_test.cc [hunk 10]
```diff
@@ -31,35 +34,69 @@ int test_metadata_rollover() {
...
   }
-#else
-  if (meta_snapshot_zones.size() == 0) {
-    std::cout << "Success! Did not allocate meta snapshot zone\n";
-    return 0;
+
+  ZenFS *zenFS;
+  zenFS = new ZenFS(zbd, FileSystem::Default(), logger);
+
+  if (FLAGS_aux_path.back() != '/') FLAGS_aux_path.append("/");
```
Comment (royguo) [inline test/zenfs_metazone_rollover_test.cc]: We still use `{}` for the single line `if` statements.

File: fs/fs_zenfs.cc [hunk 11]
```diff
@@ -1096,6 +1485,98 @@ Status ZenFS::MkFS(std::string aux_fs_path, uint32_t finish_threshold,
...
+
+  snapshot_log.reset(new ZenMetaLog(zbd_, snapshot_zone));
+
+  Superblock* snapshot_super = new Superblock(zbd_, aux_fs_path, finish_threshold,
+                                     max_open_limit, max_active_limit);
+  std::string snapshot_super_string;
+  snapshot_super->EncodeTo(&snapshot_super_string);
+
+  // write an empty superblock to snapshot zone
+  s = snapshot_log->AddRecord(snapshot_super_string);
```
Comment (royguo) [inline fs/fs_zenfs.cc]: The logic here could be merged with the previous for loop, we can directly change the original MkFS and extract the `metadata` zone for loop to a new helper function, this new help function could take parameter to persistent both op log zone and snapshot zone

File: fs/fs_zenfs.cc [hunk 12]
```diff
@@ -1016,8 +1155,258 @@ Status ZenFS::Mount(bool readonly) {
...
+  }
+
+  LogFiles();
+
+  return Status::OK();
+}
+
+/* Mount the filesystem by recovering form the latest valid snapshot zone
+ * and metadata zone*/
+Status ZenFS::MountV2(bool readonly) {
```
Comment (royguo) [inline fs/fs_zenfs.cc]: Same as `MkfsV2`, please reuse original function (you can take some logic out and create new function to handle the differences)

File: fs/fs_zenfs.cc [hunk 13]
```diff
@@ -1016,8 +1155,258 @@ Status ZenFS::Mount(bool readonly) {
...
+
+  /* We need a minimum of two non-offline snapshot zones */
+  if (snapshot_zones.size() < 2) {
+    Error(logger_,
+          "Need at least two non-offline snapshot zones to open for write");
+    return Status::NotSupported();
+  }
+
+  /* We need a minimum of two non-offline meta data zones */
+  if (metazones.size() < 2) {
```
Comment (royguo) [inline fs/fs_zenfs.cc]: maybe `iff (snapshot_zones.size() < 2 || metazones.size() < 2 )` is simpler?

File: fs/fs_zenfs.h (hunk 8 above)
Comment (weilewei) [reply fs/fs_zenfs.h]: Yes, the comment is in the function definition in `fs_zenfs.cc` file

File: fs/fs_zenfs.cc [hunk 14]
```diff
@@ -788,6 +788,23 @@ Status ZenFS::DecodeSnapshotFrom(Slice* input) {
   return Status::OK();
 }
 
+/* This function only read through slices in a snapshot record and
+ * DO NOT add valid slices to files_ to avoid duplicating files
+ * appeared in different snapshots*/
+Status ZenFS::DecodeSnapshotFromAndNoCacheFiles(Slice* input) {
```
Comment (royguo) [inline fs/fs_zenfs.cc]: The function signature looks not straightforward to me, and could you please explain a little bit more about why we need this?

Comment (royguo) [reply fs/fs_zenfs.cc]: Seems this function is only used to validate if the target slice is a valid snapshot, right? If so, please correct the name to `IsValidSnapshot` or something similar.

The original `DecodeSnapshotFrom` function changed the `files_` container so it is OK to be called `DecodeSnapShot`.

Besides that, you should modify the old `DecodeSnapshotFrom` to support your new requirement, not simply create a new function.

File: test/zenfs_metazone_rollover_test.cc [hunk 15]
```diff
@@ -31,35 +34,71 @@ int test_metadata_rollover() {
...
 
-int test_snapshot_zone_allocation() {
-  std::shared_ptr<Logger> logger;
+int test_mkfs() {
   Status s;
+  DIR* aux_dir;
+  std::shared_ptr<Logger> logger;
+
+  if (FLAGS_aux_path.empty()) {
+      fprintf(stderr, "You need to specify --aux_path\n");
```
Comment (royguo) [inline test/zenfs_metazone_rollover_test.cc]: The indents here should be 2 spaces, not 4, please format this file.

File: fs/fs_zenfs.cc [hunk 16]
```diff
@@ -823,6 +840,128 @@ Status ZenFS::DecodeFileDeletionFrom(Slice* input) {
   return Status::OK();
 }
 
+/* find the latest valid snapshot record */
+Status ZenFS::RecoverFromSnapshotZone(ZenMetaLog* log) {
```
Comment (royguo) [inline fs/fs_zenfs.cc:828]: Please merge these three functions: RecoverFromSnapshotZone, RecoverFromOpLogZone and RecoverFrom, at least they should use the same entry function.

File: fs/fs_zenfs.cc [hunk 17]
```diff
@@ -928,78 +1065,181 @@ Status ZenFS::Mount(bool readonly) {
...
+  Status s;
+
+#ifdef WITH_ZENFS_ASYNC_METAZONE_ROLLOVER
+  std::vector<Zone*> snapshot_zones = zbd_->GetSnapshotZones();
+  std::vector<std::unique_ptr<Superblock>> valid_superblocks_snapshot;
+  std::vector<std::unique_ptr<ZenMetaLog>> valid_logs_snapshot;
+  std::vector<Zone*> valid_zones_snapshot;
+  std::vector<std::pair<uint32_t, uint32_t>> seq_map_snapshot;
+
+  s = FindAllValidSuperblocks(snapshot_zones, valid_superblocks_snapshot,
```
Comment (royguo) [inline fs/fs_zenfs.cc]: Seems `snapshot_zones` is not used outside `FindAllValidSuperblocks`, I think we could put it inside to reduce the parameter number....

Comment (weilewei) [reply fs/fs_zenfs.cc]: Sure, I added `ZenFSZoneTag` enum parameter in `FindAllValidSuperblocks()` to differentiate the usage either snapshot or op log zone. So we can put the zones usage inside the `FindAllValidSuperblocks()`

File: fs/fs_zenfs.cc (hunk 16 above)
Comment (weilewei) [reply fs/fs_zenfs.cc:828]: Unfortunately, it is very difficult to merge three functions into one. since their logics are completely different, especially the 1) different functions require different amount of switch cases; 2) `break` or `continue` conditions vary in three different functions.

File: test/zenfs_metazone_rollover_test.cc (hunk 15 above)
Comment (weilewei) [reply test/zenfs_metazone_rollover_test.cc]: done.

File: fs/fs_zenfs.cc (hunk 14 above)
Comment (weilewei) [reply fs/fs_zenfs.cc]: Sure, I added a boolean `cache_files` in the `DecodeSnapshotFrom()`, so we can either 1) only validating the snapshot or 2) both validating snapshot and cache the files.

File: fs/fs_zenfs.cc [hunk 18]
```diff
@@ -1063,34 +1241,44 @@ Status ZenFS::MkFS(std::string aux_fs_path, uint32_t finish_threshold,
...
+                finish_threshold, max_open_limit, max_active_limit);
 
-  s = log->AddRecord(super_string);
-  if (!s.ok()) return std::move(s);
+  if (!s.ok()) {
+    Error(logger_, "Failed to reset op log: %s", s.ToString().c_str());
+    return Status::IOError("Failed to reset op log");
+  }
 
+#if !defined(WITH_ZENFS_ASYNC_METAZONE_ROLLOVER)
```
Comment (royguo) [inline fs/fs_zenfs.cc]: We prefer `ifndef` here

File: fs/fs_zenfs.cc [hunk 19]
```diff
@@ -1031,12 +1174,47 @@ Status ZenFS::Mount(bool readonly) {
   return Status::OK();
 }
 
+Status ZenFS::ResetZone(std::vector<Zone*> const& zones, Zone* reset_zone,
```
Comment (royguo) [inline fs/fs_zenfs.cc]: The function signature is a little bit confusing, we may need to change it to `InitMetaZone` since the code here doesn't only do the resetting but also initialize an empty super block.

File: fs/fs_zenfs.cc [hunk 20]
```diff
@@ -1031,12 +1174,47 @@ Status ZenFS::Mount(bool readonly) {
...
+    }
+  }
+
+  if (!reset_zone) {
+    return Status::IOError("No available zones\n");
+  }
+
+  log->reset(new ZenMetaLog(zbd_, reset_zone));
+
+  Superblock* super = new Superblock(zbd_, aux_fs_path, finish_threshold,
```
Comment (royguo) [inline fs/fs_zenfs.cc:1203]: The function call passed to many params, so let's see if we can move the `creation` of the superblock outside this function? And also the `reset_zone` seems not needed.... so the passed in params should be (zones, log, superblock) which is much easier to understand and maintance.

Comment (weilewei) [reply fs/fs_zenfs.cc:1203]: good point! Thanks, and I moved the creation step outside of this function. Also simplify the function signatures.

**Review by weilewei — COMMENTED** [review]
@royguo can u please take another look? Thanks.

**Comment by royguo** [comment]
All commits are included in PR49
